// @speckle/shared/dist/esm/core/utils/localStorage.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _FakeStorage_internalStorage;
function checkLocalStorageAvailability() {
    try {
        const testKey = '___localStorageAvailabilityTest';
        const storage = globalThis.localStorage;
        storage.setItem(testKey, testKey);
        storage.getItem(testKey);
        storage.removeItem(testKey);
        return true;
    }
    catch (e) {
        return false;
    }
}
/**
 * In memory implementation of the Storage interface, for use when LocalStorage
 * isn't available
 */
class FakeStorage {
    constructor() {
        _FakeStorage_internalStorage.set(this, new Map());
    }
    clear() {
        __classPrivateFieldGet(this, _FakeStorage_internalStorage, "f").clear();
    }
    getItem(key) {
        return __classPrivateFieldGet(this, _FakeStorage_internalStorage, "f").get(key) || null;
    }
    key(index) {
        return [...__classPrivateFieldGet(this, _FakeStorage_internalStorage, "f").keys()][index] || null;
    }
    removeItem(key) {
        __classPrivateFieldGet(this, _FakeStorage_internalStorage, "f").delete(key);
    }
    setItem(key, value) {
        __classPrivateFieldGet(this, _FakeStorage_internalStorage, "f").set(key, value);
    }
    get length() {
        return __classPrivateFieldGet(this, _FakeStorage_internalStorage, "f").size;
    }
}
_FakeStorage_internalStorage = new WeakMap();
/**
 * Whether or not the local storage is available in this session
 */
const isLocalStorageAvailable = checkLocalStorageAvailability();
/**
 * Localstorage (real or faked) to use in this session
 */
const internalStorage = isLocalStorageAvailable
    ? globalThis.localStorage
    : new FakeStorage();
/**
 * Utility for nicer reads/writes from/to LocalStorage without having to worry about the browser
 * throwing a hissy fit because the page is opened in Incognito mode or node not having localStorage at all
 */
export const SafeLocalStorage = {
    get(key) {
        return internalStorage.getItem(key);
    },
    set(key, value) {
        internalStorage.setItem(key, value);
    },
    remove(key) {
        internalStorage.removeItem(key);
    },
    /**
     * Flag for telling if we're using a real localStorage or faking it with a basic in-memory collection
     */
    isRealLocalStorage: isLocalStorageAvailable
};
