// @speckle/shared/dist/commonjs/environment/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFeatureFlags = void 0;
const znv_1 = require("znv");
const zod_1 = require("zod");
function parseFeatureFlags() {
    //INFO
    // As a convention all feature flags should be prefixed with a FF_
    return (0, znv_1.parseEnv)(process.env, {
        // Enables the automate module.
        FF_AUTOMATE_MODULE_ENABLED: {
            schema: zod_1.z.boolean(),
            defaults: { production: false, _: true }
        },
        // Enables the gendo ai integration
        FF_GENDOAI_MODULE_ENABLED: {
            schema: zod_1.z.boolean(),
            defaults: { production: false, _: true }
        },
        // Enables the workspaces module
        FF_WORKSPACES_MODULE_ENABLED: {
            schema: zod_1.z.boolean(),
            defaults: { production: false, _: true }
        },
        // Enables the multiple emails module
        FF_MULTIPLE_EMAILS_MODULE_ENABLED: {
            schema: zod_1.z.boolean(),
            defaults: { production: false, _: true }
        },
        // Disables writing to the closure table in the create objects batched services (re object upload routes)
        FF_NO_CLOSURE_WRITES: {
            schema: zod_1.z.boolean(),
            defaults: { production: false, _: false }
        }
    });
}
let parsedFlags;
function getFeatureFlags() {
    if (!parsedFlags)
        parsedFlags = parseFeatureFlags();
    return parsedFlags;
}
exports.getFeatureFlags = getFeatureFlags;
